### PR TITLE
Deduplicate transactions in pending blocktree

### PR DESCRIPTION
**Pull request by abenedito-monad**
Provide TransactionList of each pending block in the extending branch to the mempool on FetchTxs so that it can avoid these transactions and return unused transactions.

### CODE REVIEW

File: monad-consensus-state/src/lib.rs [hunk 1]
```diff
@@ -654,30 +656,26 @@ where
...
                 None => ConsensusAction::ProposeEmpty,
-            }
-        } else {
-            ConsensusAction::Abstain
+            };
         }
+
+        if self.config.propose_with_missing_blocks {
+            return ConsensusAction::ProposeEmpty;
+        };
```
Comment (ariqchowdhury) [inline monad-consensus-state/src/lib.rs]: the config name is probably confusing, but the intent of `propose_with_missing_blocks` is to allow a node to make a full proposal even if its blocktree is not complete.
The `ProposeEmpty` action is for the case where `get_next_state_root` fails.

File: monad-consensus-state/src/lib.rs [hunk 2]
```diff
@@ -654,30 +656,26 @@ where
...
+
+        ConsensusAction::Abstain
     }
 }
 
 pub enum ConsensusAction {
-    Propose(Hash),
+    Propose(Hash, Vec<TransactionList>),
```
Comment (michael-yxchen) [inline monad-consensus-state/src/lib.rs:711]: Not related this PR specifically. I think we should start thinking about the actual format/encoding of the `TransactionList`, and document it somewhere.

Comment (abenedito-monad) [reply monad-consensus-state/src/lib.rs:711]: The point of TransactionList being a `Vec<u8>` is for consensus to be completely agnostic as to the underlying type of transactions. The actual format/encoding of the TransactionList data we'll be using in testnet/mainnet will be an RLP encoding of a list of ethereum transaction hashes.

Comment (michael-yxchen) [reply monad-consensus-state/src/lib.rs:711]: oh. I understand the `Vec<Vec<u8>>` in `create_proposal` now. Thanks

File: monad-consensus-state/src/lib.rs (hunk 1 above)
Comment (abenedito-monad) [reply monad-consensus-state/src/lib.rs]: Fixed!

File: monad-consensus-state/src/lib.rs [hunk 3]
```diff
@@ -655,30 +657,30 @@ where
...
+        if let Some(pending_txs) = self.pending_block_tree.get_txs_on_path_to_root(parent_bid) {
+            return ConsensusAction::Propose(h, pending_txs);
         }
+
+        // Still propose but with the chance of proposing duplicate txs
+        if self.config.propose_with_missing_blocks {
+            return ConsensusAction::Propose(h, vec![]);
+        };
+
+        ConsensusAction::Abstain
```
Comment (michael-yxchen) [inline monad-consensus-state/src/lib.rs:694]: I think it's changing the outcome of one particular case. When we're missing blocks to the root, and `propose_with_missing_blocks == false`, the original code outputs `Abstain` and the PR outputs `ProposeEmpty`.

What's the reason for this change?

Comment (ariqchowdhury) [reply monad-consensus-state/src/lib.rs:694]: andre and I talked about this offline before changing it, I think that it makes sense to propose empty in this case as its similar to the case of missing state_root_hashes. Basically, the node is live but missing "something" that allows it to properly propose, so it should propose empty to keep consensus on the happy path in both cases.

Comment (michael-yxchen) [reply monad-consensus-state/src/lib.rs:694]: Cool, just wanted to make sure it was an intentional change.

That's reasonable, my only concern's that if the proposal comes after a quorum of the votes, then the leader would request block sync and immediately propose an empty block. That's not a huge deal as the network layer should prioritize sending to the next leader. Just wanted to bring it up so we're aware.

The rest of the PR looks good to me.
